src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java

```diff
@@ -729,6 +733,8 @@ private static void register(Descriptor d) {
                 m(CKM_BLOWFISH_CBC));
         d(SKF, "ChaCha20",      P11SecretKeyFactory,
                 m(CKM_CHACHA20_POLY1305));
+        d(SKF, "Generic",       P11SecretKeyFactory,
+                m(CKM_GENERIC_SECRET_KEY_GEN));
```

Comment:
How useful is this? Do you intend to import a "Generic" `SecretKeySpec` into a token?

Reply:
Generic is a native PKCS11 key type (`CKK_GENERIC_SECRET`) that could have been added to SunPKCS11 before, irrespective of HKDF. It's convenient for the test to have key material in the token and test consolidation (IKM or salt).

Reply:
Do you think we can add it to the Java Security Standard Names document?

Reply:
Yes, sure. I added a comment to the _Solution_ section of the CSR and the name to the table in _Specification_.

Reply:
Oh, I meant https://docs.oracle.com/en/java/javase/23/docs/specs/security/standard-names.html#secretkeyfactory-algorithms. If you think it's worth adding there, please provide Algorithm Name and Description.

Reply:
Hi @wangweij,

Which test have you executed? I'm able to use "Generic" keys for HmacSHA256, in a local slowdebug build of this branch. 

```properties
cat >providersList.properties <<'EOF'
security.provider.1=SunPKCS11 --\\n\
    name = NSS\\n\
    nssLibraryDirectory = /usr/lib64\\n\
    nssDbMode = noDb
security.provider.2=SUN
security.provider.3=SunRsaSign
security.provider.4=SunEC
security.provider.5=SunJSSE
security.provider.6=SunJCE
security.provider.7=SunJGSS
security.provider.8=SunSASL
security.provider.9=XMLDSig
security.provider.10=SunPCSC
security.provider.11=JdkLDAP
security.provider.12=JdkSASL
EOF
```

```java
cat >Main.java <<'EOF'
import java.util.HexFormat;
import javax.crypto.Mac;
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.SecretKeySpec;

public final class Main {
    public static void main(String[] args) throws Exception {
        byte [] keyMaterial = "Secret-Bytes".getBytes();
        SecretKeySpec spec = new SecretKeySpec(keyMaterial, "Generic");
        SecretKeyFactory skf = SecretKeyFactory.getInstance("Generic");
        SecretKey sk = skf.generateSecret(spec);
        System.out.println(sk);

        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(sk);
        mac.update("test".getBytes());
        System.out.println(HexFormat.of().formatHex(mac.doFinal()));
    }
}
EOF
```

```bash
./build/linux-x86_64-server-slowdebug/images/jdk/bin/java \
    -Djava.security.properties=providersList.properties Main.java
rm providersList.properties Main.java
```

Output:
```
SunPKCS11-NSS Generic secret key, 96 bits session object, not sensitive, extractable)
c5dca603b87a1a1fe264f3cab2f851d513afdd2a7dd5ed3ee337356e2d7a001a
```

Reply:
The key has to have `CKA_SIGN = true` in order to be used for a HMAC operation. For example, you can modify the code snippet shared by @franferrax to include the line `Mac mac = Mac.getInstance("HmacSHA256", "SunPKCS11-NSS");` in _Main.java_ (instead of `Mac mac = Mac.getInstance("HmacSHA256");`) and the line `attributes = compatibility` in _providersList.properties_. With these changes, I get the following output:

```
./bin/java -Djava.security.properties=providersList.properties Main.java
SunPKCS11-NSS Generic secret key, 96 bits session object, not sensitive, extractable)
c5dca603b87a1a1fe264f3cab2f851d513afdd2a7dd5ed3ee337356e2d7a001a
```

Reply:
Yes, it works now with the `attributes = compatibility` line. Told you I am not an expert. Thanks.

Reply:
Hi, sorry for the confusion. Even when _SunPKCS11_ was the first provider in the list, I should have checked what provider was implementing the `Mac`.

In my previous example, due to _delayed provider selection_, the provider is chosen during the `mac.init(sk)` call. _SunPKCS11_ throws the `InvalidKeyException` you were reproducing (caused by `PKCS11Exception: CKR_KEY_TYPE_INCONSISTENT`), but this exception is ignored and _SunJCE_ is finally selected. This doesn't happen if `attributes = compatibility` or `attributes(*,CKO_SECRET_KEY,CKK_GENERIC_SECRET)={ CKA_SIGN=true }` is used.